### ISSUE TITLE
Bug: The rule 'no-unused-vars' works incorrectly with PUG template in Vue 3 (script+setup)

### ISSUE THREAD

**Issue by tpuida** [eslint/eslint#18096]
### Environment

Node version: v21.6.1
npm version: 10.2.4
Local ESLint version: v8.44.0
Global ESLint version: v8.56.0
Operating System: MacOS 14

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```
/* eslint-env node */
require('@rushstack/eslint-patch/modern-module-resolution');

module.exports = {
    root: true,
    extends: [
        'plugin:vue/vue3-essential',
        'eslint:recommended',
        '@vue/eslint-config-typescript',
        '@vue/eslint-config-prettier/skip-formatting',

        '@vue/eslint-config-airbnb-with-typescript',
    ],
    rules: {
        indent: ['error', 4, { SwitchCase: 1 }],
        'no-console': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
        'no-debugger': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
        'no-extra-boolean-cast': 'off',
        'no-unused-vars': [
            'warn',
            {
                argsIgnorePattern: '^_',
                varsIgnorePattern: '^_',
                caughtErrorsIgnorePattern: '^_',
                ignoreRestSiblings: true,
            },
        ],
        '@typescript-eslint/no-unused-vars': 'off',
        'vue/multi-word-component-names': process.env.NODE_ENV === 'production' ? 'warn' : 'off',
        'vue/order-in-components': [
            'error',
            {
                order: [
                    'el',
                    'name',
                    'functional',
                    ['delimiters', 'comments'],
                    ['components', 'directives', 'filters'],
                    'extends',
                    'mixins',
                    ['provide', 'inject'],
                    'validate',
                    'inheritAttrs',
                    'model',
                    'emits',
                    'ROUTER_GUARDS',
                    ['props', 'propsData'],
                    'setup',
                    'asyncData',
                    'data',
                    'validations',
                    'computed',
                    'watch',
                    'watchQuery',
                    'apollo',
                    'LIFECYCLE_HOOKS',
                    'methods',
                    ['template', 'render'],
                    'renderError',
                ],
            },
        ],
        'max-len': [
            'warn',
            {
                code: 150,
                ignorePattern: 'd="([\\s\\S]*?)"',
                ignoreComments: true,
            },
        ],
        'vue/max-len': 'off',
        'object-curly-newline': [
            'error',
            {
                ObjectExpression: {
                    consistent: true,
                },
                ObjectPattern: {
                    multiline: true,
                },
                ExportDeclaration: {
                    multiline: true,
                    minProperties: 3,
                },
            },
        ],
        'no-param-reassign': [
            'error',
            {
                props: false,
            },
        ],
        camelcase: 'off',
        'no-plusplus': 'off',
        'no-restricted-syntax': 'off',
        'no-underscore-dangle': ['error', { allow: ['__typename'] }],
        'import/extensions': [
            'error',
            'ignorePackages',
            {
                ts: 'never',
            },
        ],
        'import/no-cycle': 'off',
        'import/prefer-default-export': 'off',
        'prefer-destructuring': [
            'error',
            {
                object: true,
                array: false,
            },
        ],
        'operator-linebreak': 'off',
        'implicit-arrow-linebreak': 'off',
        'function-paren-newline': 'off',
        'no-confusing-arrow': 'off',
        'vue/block-lang': [
            'error',
            {
                script: {
                    lang: ['ts', 'js'],
                },
            },
        ],
        '@typescript-eslint/indent': 'off',
        '@typescript-eslint/lines-between-class-members': 'off',
    },
    overrides: [
        {
            files: ['cypress/e2e/**/*.{cy,spec}.{js,ts,jsx,tsx}'],
            extends: ['plugin:cypress/recommended'],
        },
    ],
    parserOptions: {
        ecmaVersion: '2022',
        tsconfigRootDir: __dirname,
        files: ['**/*.ts', '**/*.vue'],
    },
};

```
</details>
it works well:

```vue
<template>
  <div class="image-preview">
    <img :src="imageUrl" alt="" />
  </div>
</template>

<script lang="ts" setup>
import { ref } from 'vue/dist/vue.js';

const imageUrl = ref('');
</script>
```
with warning
```vue
<template lang="pug">
.image-preview
    img(:src="imageUrl" alt="")
</template>

<script lang="ts" setup>
import { ref } from 'vue/dist/vue.js';

const imageUrl = ref(''); // warning: 'imageUrl' is assigned a value but never used. Allowed unused vars must match /^_/u  no-unused-vars
</script>

```

### What did you expect to happen?

no warning 

### What actually happened?

` 9:7  warning  'imageUrl' is assigned a value but never used. Allowed unused vars must match /^_/u  no-unused-vars`

### Link to Minimal Reproducible Example

//

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

I faced this problem only in the composition API of vue 3 (script+setup) with the PUG template. Optional API with PUG doesn't show this warning

**Comment by Rec0iL99** [comment]
Hi @tpuida, thanks for the issue. I don't think this is a bug because the variable name `imageUrl` does not match regexp provided for the option `varsIgnorePattern`. The reqexp only matches variable names that start with an `_`.

```vue
/*eslint no-unused-vars: ["error", { "varsIgnorePattern": "^_" }]*/

<template lang="pug">
.image-preview
    img(:src="imageUrl" alt="")
</template>

<script lang="ts" setup>
import { ref } from 'vue/dist/vue.js';

const _imageUrl = ref(''); // no-unused-vars will now not complain here 
</script>
```

[playground](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHsgXCJ2YXJzSWdub3JlUGF0dGVyblwiOiBcIl5fXCIgfV0qL1xuaW1wb3J0IHsgcmVmIH0gZnJvbSAndnVlL2Rpc3QvdnVlLmpzJztcblxuY29uc3QgX2ltYWdlVXJsID0gcmVmKCcnKTsiLCJvcHRpb25zIjp7ImVudiI6e30sInJ1bGVzIjp7fSwicGFyc2VyT3B0aW9ucyI6eyJlY21hRmVhdHVyZXMiOnt9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)

**Comment by tpuida** [comment]
Thanks for your quick response. The problem is that this variable is used in the template. Here is another short working example
```vue
<template lang="pug">
button(@click="handleClick") clicked {{ count }} times
</template>

<script lang="ts" setup>
import { ref } from 'vue';

const count = ref(0);
const handleClick = () => { // 'handleClick' is assigned a value but never used. Allowed unused vars must match /^_/u.eslint[no-unused-vars](https://eslint.org/docs/latest/rules/no-unused-vars)
    count.value++;
};
</script>
```
You can see in the demo that `handleClick` is used
[demo](https://play.vuejs.org/#eNp9UE1PAjEQ/StNL0Igiwk3XIhKOOhBjXrsZS3DUui2TTtFks3+d6ddQQ6GTTZp38fM62v5g3PFIQKf8RKhcbpCWAhTfkVEa9i91Eru54JvK7PWsEw3wRcsw7BmbcukjQZZ1zFUDYRy0jvTjMnFQLoG6ZVDpitT00AMgrMAGB2xqnHWI2uZhw3r2Mbbht1Qqpu75JTWBPzdM0+Swe2QiB6+CEbkYMjmC9YKw+jLjuJQ6QijERk6+stJH4OW8jGFoCEbVRe7YA1VkI2CS9s4pcG/OlS0RPDZaaTgldb2+zlj6COMT7jcgtz/g+/CMWGCv3kI4A8g+JnDyteAPb36eIEjnc9kY9dRk/oK+Q7B6pgy9rLHmLrwF7qc9im3q0z9GVZHBBNOj0pBk7LLesGp8eWVp//FnRbT7KNOefcDtxXB3g==)

**Comment by snitin315** [comment]
@tpuida By default, eslint is unable to parse vue templates ([Demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyplc2xpbnQgbm8tdW51c2VkLXZhcnM6IFtcImVycm9yXCIsIHsgXCJ2YXJzSWdub3JlUGF0dGVyblwiOiBcIl5fXCIgfV0qL1xuXG48dGVtcGxhdGU+XG4gIDxkaXYgY2xhc3M9XCJpbWFnZS1wcmV2aWV3XCI+XG4gICAgPGltZyA6c3JjPVwiaW1hZ2VVcmxcIiBhbHQ9XCJcIiAvPlxuICA8L2Rpdj5cbjwvdGVtcGxhdGU+XG5cbjxzY3JpcHQgbGFuZz1cInRzXCIgc2V0dXA+XG5pbXBvcnQgeyByZWYgfSBmcm9tICd2dWUvZGlzdC92dWUuanMnO1xuXG5jb25zdCBpbWFnZVVybCA9IHJlZignJyk7XG48L3NjcmlwdD5cbiIsIm9wdGlvbnMiOnsiZW52Ijp7fSwicnVsZXMiOnt9LCJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9LCJlY21hVmVyc2lvbiI6ImxhdGVzdCIsInNvdXJjZVR5cGUiOiJtb2R1bGUifX19)).

It seems you are using [eslint-plugin-vue](https://eslint.vuejs.org/) which enables linting vue templates, so you should report this issue in the https://github.com/vuejs/eslint-plugin-vue repository. I don't think we can help you here. You may want to turn off `no-unused-var` in favor of [vue/no-unused-var](https://eslint.vuejs.org/rules/no-unused-vars.html) for vue templates.
